### ISSUE TITLE
Hostnames for localhost  HTTP gateway

### ISSUE THREAD

**Issue by lidel** [ipfs/in-web-browsers#109]
## Motivation

- At some point  in the future we may have native protocol handler ([libdweb](https://github.com/mozilla/libdweb/#protocol-api)+https://github.com/ipfs-shipyard/ipfs-companion/pull/533 or similar), but for now the URL in address bar remains <del>fugly</del> less than pretty
- The latest [Companion](https://github.com/ipfs/ipfs-companion/) extension supports [DNSLink](https://docs.ipfs.io/guides/concepts/)-based  redirects from URLs like `http://docs.ipfs.io` to a local go-ipfs gateway at  `http://127.0.0.1:8080/ipns/docs.ipfs.io/`
- Our public gateways start to support cidv1b32 in Subdomains (https://github.com/ipfs/in-web-browsers/issues/89), which restores Origin-based isolation between websites. This also means we don't want to redirect such URLs to local gateway as it would remove this isolation.

## Requirements

I wonder if there is a way of generating vanity hostnames pointing at localhost that:
- [ ] does not require lookups to an external DNS server
- [ ] keeps HTML content in [secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts)
- [ ] works out-of-the box on windows, mac and linux
     - alternatively we can add heuristics to pick method based on user's platform
- [ ] lets us introduce [subdomain-driven Origin isolation](https://github.com/ipfs/in-web-browsers/issues/89) for locally running gateway
- [ ] (nice to have, but we can work around this for embedds) does not trigger mixed-content warnings when IPFS resource from localhost is embedded on HTTPS website
   - it is what happens when `/ipfs/{cid}` is redirected from public gateway running on https to a custom one running on http
   - <del>this may be the blocker: AFAIK only `127.0.0.1` and `::1` are whitelisted by browser vendors (eg. [firefox #903966](https://bugzilla.mozilla.org/show_bug.cgi?id=903966)) but `localhost` is not (that is why Companion uses raw IP)</del> 
    - as of 2019-10 `localhost` is a secure context in both Firefox and Chrome. Firefox does not support subdomains (`foo.bar.localhost` is not a secure context)

If anyone has unconventional ideas, feel free to comment.

# Candidates

### Current status

:lock: – `window.isSecureContext === true`
:anger: – `window.isSecureContext === false`

| Status | Firefox | Chrome |
| ---- | ---- | ---- |
| `127.0.0.1` | :lock: | :lock: |
| `localhost` | :lock: | :lock: |
| `foo.bar.localhost` | :anger: | :lock: |

### References 

- https://w3c.github.io/webappsec-secure-contexts/#localhost:
  >  Section 6.3 of [RFC6761] lays out the resolution of localhost. and names falling within .localhost. as special, and suggests that local resolvers SHOULD/MAY treat them specially. For better or worse, resolvers often ignore these suggestions, and will send localhost to the network for resolution in a number of circumstances.
- [Let 'localhost' be localhost](https://tools.ietf.org/html/draft-ietf-dnsop-let-localhost-be-localhost)
  - [[DNSOP] DNSOP Call for Adoption - draft-west-let-localhost-be-localhost](https://mailarchive.ietf.org/arch/msg/dnsop/eGn3kxRhoj1r4c-zHbE1_ozKq_k)
- bugzilla:
  - [Stop treating 'localhost' as securely delivered for the purposes of Secure Contexts](https://bugzilla.mozilla.org/show_bug.cgi?id=1346835)
  - [Consider hardcoding localhost names to the loopback address](https://bugzilla.mozilla.org/show_bug.cgi?id=1220810#c23)
  - [Stop treating "http://localhost/" (by name) as mixed content](https://bugzilla.mozilla.org/show_bug.cgi?id=1488740)
- chromium
  - [Intent to Implement and Ship: Treat `http://localhost` as a secure context](https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/RC9dSw-O3fE/discussion)
  - [Issue 589141: *.localhost should be considered secure](https://bugs.chromium.org/p/chromium/issues/detail?id=589141)
- go-ipfs
  - [localhost subdomains do not work on Firefox and Safari](https://github.com/ipfs/go-ipfs/issues/7527)

**Comment by lidel** [comment]
Candidate:  `http://{fqdn}.ipns.localhost`
Test URL: http://docs.ipfs.io.ipns.localhost:8080/ipns/docs.ipfs.io
Result: :-1:  (Works on a box with unbound, does not work on stock Debian)

**Comment by olizilla** [comment]
There are some occult tricks for browsers on mac and windows; you can do some automagic with this https://en.wikipedia.org/wiki/Proxy_auto-config 

I found it via https://github.com/KishanBagaria/frwrd

**Comment by lidel** [comment]
> [..] you can do some automagic with this https://en.wikipedia.org/wiki/Proxy_auto-config

Changing proxy setting globallyfor the user is not a good idea, however Firefox provides [browser.proxy API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/proxy) with a very interesting property:

> use the [`proxy.onRequest`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/proxy/onRequest) event listener to intercept web requests, and return an object that describes whether and how to proxy them.

This would enable us to use PAC-like processing only for IPFS requests.
(Right now we just redirect them)

**Comment by lockedshadow** [comment]
>  Linux -1  (Works on a box with unbound, does not resolve on stock Debian)

Seems like that it also depends on `myhostname` plugin for `/etc/nsswtich.conf`, which is not installed by default in Debian. With this plugin installed and activated, the test URL actually redirects to `http://{local_gateway}/ipns/docs.ipfs.io/`.

From `man nss-myhostname`:
> The hostnames "localhost" and "localhost.localdomain" (as well as any hostname ending in ".localhost" or ".localhost.localdomain") are resolved to the IP addresses 127.0.0.1 and ::1.

**Comment by lidel** [comment]
I think switching from `127.0.0.1` to `::1` where possible is the best we can do right now, given cross-platform constraints.

I've set up local gateway at dual stack and `::1` looks less intimidating:

> ![2019-01-08-125329_758x575_scrot](https://user-images.githubusercontent.com/157609/50836770-5248d500-135a-11e9-924c-4474328af23f.png)

 I created a proposal to extend default `Addresses.Gateway` with `/ip6/::1/tcp/8080` – see  https://github.com/ipfs/go-ipfs/issues/5905

**Comment by lidel** [comment]
Update: there may be a way to support `http://<cidv1b32>.ipfs.localhost` on every platform supporting HTTP Proxies – see proposal in https://github.com/ipfs/go-ipfs/issues/5982

**Comment by lidel** [comment]
Notes on Secure Context and `.ipfs.localhost` – https://github.com/ipfs-shipyard/ipfs-companion/issues/667#issuecomment-467669949

**Comment by lidel** [comment]
The ".dweb" Special-Use Domain Name:  https://github.com/arewedistributedyet/arewedistributedyet/issues/34

**Comment by lidel** [comment]
Interesting development:
As of 2019-10 `localhost` (hostname) is a secure context in both Firefox and Chrome. 

Caveat: Firefox does not include subdomains (`foo.bar.localhost` is not a secure context), but I posted our use case in related bug to match behavior from Chrome: 
https://bugzilla.mozilla.org/show_bug.cgi?id=1433933#c6

There is also a discussion to skip DNS lookup for `*.localhost` (https://bugzilla.mozilla.org/show_bug.cgi?id=1220810) and hardcode it to loopback, which would make it work on all platforms, regardless of local DNS setup. 

### Demo

Loaded http://en.wikipedia-on-ipfs.org.ipfs.localhost:8080/ipns/en.wikipedia-on-ipfs.org/wiki/ today and Chrome marks it as secure context:

| Firefox 69 | Chrome 78 | 
| ---- | ---- | 
| isSecureContext: false | isSecureContext: true |
| ![secure-context-firefox-2019-10-01--15-30-25](https://user-images.githubusercontent.com/157609/65966704-c6835580-e460-11e9-91e3-c0b91583d315.png) | ![secure-context-chrome-2019-10-01--15-30-25](https://user-images.githubusercontent.com/157609/65966709-cb480980-e460-11e9-8a84-d664dd229237.png) |

**Comment by lidel** [comment]
Quick update:

Since go-ipfs 0.5, when user runs local IPFS node, they can load websites and content via subdomains like this:

http://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.ipfs.localhost:8080/wiki/
http://bafybeid6grcjmnvsdaeklremt2lcqk7jevbjtwrerzgq46nzxfgn5isrne.ipfs.localhost:8080/images/ipfs-logo.svg

There are two known problems with `*.localhost` at the moment:

1. If browser vendor does not hardcode `localhost` to point at loopback, then it is resolved by OS DNS resolver, and on some platforms that can produce error. Fix here is to implement section 6.3.3 from https://tools.ietf.org/html/rfc6761#section-6.3

2. Content at `*.localhost` should be marked as Secure Context. Right now some vendors such as Firefox handle `localhost` and `*.localhost` differently, partially due to legacy code, partially due to my previous point.

Implementing "let localhost be localhost" (6.3.3 from rfc6761) would ensure localhost can resolve on every platform and can be marked as secure context – very likely solving both issues.


Firefox/Gecko already is tracking this as part of https://bugzilla.mozilla.org/show_bug.cgi?id=1220810.

Saferi/Webkit: https://bugs.webkit.org/show_bug.cgi?id=171934.


As illustrated by linked issues, vendors are mostly ok with this change, but it is a low priority and unfortunately the movement on Safari and Firefox fronts to resolve it turned out to be slower than we expected. Looking into ways we can accelerate that.

**Comment by lidel** [comment]
Good news: Chromium officially locked `*.localhost` to loopback as per https://tools.ietf.org/html/draft-ietf-dnsop-let-localhost-be-localhost-02 :partying_face: 


Details from  https://bugs.chromium.org/p/chromium/issues/detail?id=589141#c15:
> We addressed this in https://bugs.chromium.org/p/chromium/issues/detail?id=691930 but didn't close this bug. There, we locked `*.localhost` to loopback as per https://tools.ietf.org/html/draft-ietf-dnsop-let-localhost-be-localhost-02 (see https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_resolver_manager.cc;drc=25292af6cee762330490f47e79e09a071dc4b5a9;l=3091). This means that we also consider `*.localhost` secure (see https://source.chromium.org/chromium/chromium/src/+/master:services/network/public/cpp/is_potentially_trustworthy.cc;drc=c70af83bc44f6829277cdc3621e7015d6e0d7530;l=228).

This increases likelihood of other vendors to follow.

**Comment by ianopolous** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1220810#c89

**Comment by autonome** [comment]
merges to mozilla-central - can't trust them until 24hrs old ;)

**Comment by lidel** [comment]
Some updates and reading for anyone interested in following `*.localhost` space :-) 

- Chromium
  - [Intent to Implement and Ship: Treat `http://localhost` as a secure context.](https://groups.google.com/a/chromium.org/g/blink-dev/c/RC9dSw-O3fE/m/sG3k73YXBQAJ)
- Firefox 
  - [Intent to prototype & ship: Treat localhost addresses as "Potentially Trustworthy"](https://groups.google.com/u/1/g/mozilla.dev.platform/c/sZdEYTiEBdE)
  - "Hardcode localhost to loopback" patch landed last week, sounds like it is here to stay :partying_face: 
    https://bugzilla.mozilla.org/show_bug.cgi?id=1220810 → https://phabricator.services.mozilla.com/D92716
    - some remaining fixes need to land before its 100% hardcoded. Work tracked in:
      https://bugzilla.mozilla.org/show_bug.cgi?id=1673364 → https://phabricator.services.mozilla.com/D94726
  - Tests to ensure `localhost` and `*.localhost` can actually not be overridden by non-local IP:
    https://bugzilla.mozilla.org/show_bug.cgi?id=1673315 →  https://phabricator.services.mozilla.com/D94712   
  -  Create a WPT test to verify handling of `locahost` and `*.locahost`
     https://bugzilla.mozilla.org/show_bug.cgi?id=1672323

**Comment by lidel** [comment]
This was resolved in https://github.com/ipfs/kubo/pull/6096

We went with `*.ipfs.localhost`  and `*.ipns.localhost`, as it works nearly everywhere: thanks to collaboration with Igalia we now have secure context in both Chromium-based browsers (Edge, Chrome, Brave, Opera), and Firefox.

Webkit/Safari  issues are actually a problem with macOS DNS resolver – tracked in https://github.com/ipfs/in-web-browsers/issues/206
